password/src/Data/Password/Validate.hs

```diff
@@ -0,0 +1,237 @@
+    isValidPolicy =
+      if isValidPasswordPolicy passwordPolicy charSetPredicate
+        then Nothing
+        else Just $ InvalidPasswordPolicy passwordPolicy
+    isTooLong :: Maybe InvalidReason
+    isTooLong =
+      if T.length password <= maximumLength
+        then Nothing
+        else Just $ PasswordTooLong maximumLength (T.length password)
+    isTooShort :: Maybe InvalidReason
+    isTooShort =
+      if T.length password >= minimumLength
+          then Nothing
+          else Just $ PasswordTooShort minimumLength (T.length password)
+    isUsingPolicyCharSetPredicate :: Maybe InvalidReason
+    isUsingPolicyCharSetPredicate =
+        let filteredText = T.filter (\c -> not $ (getCharSetPredicate charSetPredicate) c) password
+        in if T.null filteredText
+          then Nothing
+          else Just $ InvalidChar filteredText
+    hasRequiredChar :: Maybe Int -> CharacterCategory -> Maybe InvalidReason
+    hasRequiredChar Nothing _ = Nothing
+    hasRequiredChar (Just requiredCharNum) characterCategory =
+      let predicate = case characterCategory of
+            Uppercase -> isAsciiUpper
+            Lowercase -> isAsciiLower
+            Special   -> isSpecial
+            Digit     -> isDigit
+          actualRequiredCharNum = T.length $ T.filter predicate password
+       in if actualRequiredCharNum >= requiredCharNum
+          then Nothing
+          else Just $ NotEnoughReqChars characterCategory requiredCharNum actualRequiredCharNum
+
+-- | Checks if given 'PasswordPolicy' is valid
+--
+-- >>> isValidPasswordPolicy defaultPasswordPolicy defaultCharSetPredicate
+-- True
+isValidPasswordPolicy :: PasswordPolicy -> CharSetPredicate -> Bool
+isValidPasswordPolicy PasswordPolicy{..} charSetPredicate =
+  and
+   [ max minimumLength sumRequiredChars <= maximumLength
+   , minimumLength > 0
+   , maximumLength > 0
+   , isPositive uppercaseChars
+   , isPositive lowercaseChars
+   , isPositive specialChars
+   , isPositive digitChars
+   , requiredCharsetValid
+   ]
+  where
+    isPositive :: Maybe Int -> Bool
+    isPositive mNum = maybe True (> 0) mNum
+    sumRequiredChars :: Int
+    sumRequiredChars =
+      sum $ fromMaybe 0 <$> [uppercaseChars, lowercaseChars, specialChars, digitChars]
+    -- Check that if PasswordPolicy states that the passwords requires certain
+    -- characters, then CharSetPredicate should return true on at least one of
+    -- the characters that is required.
+    requiredCharsetValid :: Bool
+    requiredCharsetValid =
+      let charSets = accumulateCharSet
+            [ (uppercaseChars, isAsciiUpper)
+            , (lowercaseChars, isAsciiLower)
+            , (specialChars, isSpecial)
+            , (digitChars, isDigit)
+            ]
+      in all (\str -> any (getCharSetPredicate charSetPredicate) str) charSets
+    accumulateCharSet :: [(Maybe Int, Char -> Bool)] -> [String]
+    accumulateCharSet = map (\t -> filter (snd t) defaultCharSet) . filter (isJust . fst)
```

Comment:
Fix the log which was previously broken. 
 It nowt checks that if PasswordPolicy states that the passwords requires certain characters, then CharSetPredicate should return true on at least one of the characters that is required.

Reply:
Yes, this does make more sense. It also makes sense then to use the `defaultCharSet`.

I'd maybe find it more intuitive to split the CharSetPredicate validation out of the PasswordPolicy validation. Since the function name is `isValidPasswordPolicy`. And one should be able to just validate the policy on its own, though I think it might be better still to have a `validatePasswordPolicy` function that returns `InvalidReason`s to be more precise on what's wrong with the policy.

And then the checking of the CharSetPredicate in combination with the PasswordPolicy would happen in `validatePassword` and also give a specific `InvalidReason`.